Schema evolution post (#50)
vendia/blog: posts/2021-12-15-schema-evolution.md

---
title: "Schema Evolution"
description: "Exploring the ability to update the data model in a Universal Application (Uni)"
date: "2021-12-15"
categories:
  - announcements
authors:
  - Brian McNamara
---

> Vendia Share allows users to share a single source of truth in a secure and scalable manner through Universal Applications (Unis). Unis are automatically created using a user-supplied data model. Vendia Share takes the user supplied data model and deploys a Uni in a matter of minutes.  This includes GraphQL APIs to interact with the Uni and a serverless distributed ledger that keeps data consistent across all participants - all managed as a service.

Reviewer: ```suggestion
Vendia Share allows users to share a single source of truth in a secure and scalable manner through Universal Applications (Unis). Unis are automatically created using a user-supplied data model. Vendia Share takes the user supplied data model and deploys a Uni in a matter of minutes.  This includes provisioning fully managed GraphQL APIs to interact with the Uni and a serverless distributed ledger that keeps data consistent across all participants.
```

> As Uni participants grow and as the fidelity of solutions increase over time, it's common for users to want to evolve the original data model used to create the Uni.  Historically, there was no way for users to change the data model once a Uni was created. Any data model change to a Uni required destroying it and its accompanying nodes and recreating it with the new model. 

Reviewer: ```suggestion
As Uni participants grow and as the fidelity of solutions increase over time, it's common for users to want to revise the original data model used to create the Uni.  Historically, there was no way for users to change the data model once a Uni was created. Any data model change to a Uni required destroying it and its accompanying nodes and recreating it with the new model. 
```

Reviewer: I know we call it evolution, but that term doesn't resonant with me. It could be because I don't know the space well and that's the right term. But to me revise makes more sense than evolve.

Author: It may be a semantic thing but I equate evolve with grow. This is important as schema evolution doesn't allow for any revision - only those that are more additive in nature.

> This process was not ideal for users. 

Reviewer: ```suggestion
This process was painful for users. All the data in database was lost and users would have to rebuild the database themselves. 
```

Reviewer: Maybe we can talk about why this is so painful for customers. Not sure if this is enough to drive the point home or not.

Author: I added some context - hoping it helps with the narrative.

  "type": "object",
  "properties": {
    "Product": {
      "description": "Product information",
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "description": "Product name",
            "type": "string"
          },
          "quantity": {
            "description": "Product quantity",
            "type": "integer"
          },
          "sku": {
            "description": "Product SKU",
            "type": "string"
          },
          "description": {
            "description": "Product description",
            "type": "string"
          },
          "price": {
            "description": "Product price",
            "type": "number"
          },
          "tags": {
            "description": "Product tags",
            "type": "array",
            "items": {
              "type": "string"
            }
          },
          "size": {
            "description": "Product size",
            "type": "string",
            "enum": [
              "S",
              "M",
              "L",
              "XL"
            ]
          }
        }
      }
    }
  }
}
```
</details>

> We can use the Vendia Share CLI to instruct the Share service to apply the new data model to our existing Uni.

Reviewer 2: Do we know why the "force" option is available?  Seems like the opposite of what we'd want to allow.

> We can use the Vendia Share CLI to instruct the Share service to apply the new data model to our existing Uni.

Author: I don't - I can follow-up for context but I don't think it's a blocker for the post.
